File: src/engine/asl/transforms/callTransform.ts
import type { BaseNode } from '@/system/types';
import type { ASLStatement, ASLExpr } from '../ASLTypes';
import { transformExpr } from './exprTransform';
import type { TransformContext } from './context';

export function transformCallToStmt(node: BaseNode, ctx: TransformContext): ASLStatement | null {
  const callee = node.attributes.callee;

  if (callee === 'pinMode') {
    const modeNode = node.children[1];
    let mode: 'INPUT' | 'OUTPUT' | 'INPUT_PULLUP' = 'OUTPUT';

    if (modeNode.nodeType === 'Literal') {
      const v = modeNode.attributes.value;
      if (v === 0) mode = 'INPUT';
      if (v === 2) mode = 'INPUT_PULLUP';
    }

    return {
      kind: 'pinMode',
      pin: transformExpr(node.children[0]),
      mode,
    } as ASLStatement;
  }

  if (callee === 'digitalWrite') {
    return {
      kind: 'digitalWrite',
      pin: transformExpr(node.children[0]),
      value: transformExpr(node.children[1]),
    } as ASLStatement;
  }

  if (callee === 'Serial.print' || callee === 'Serial.println') {
    return {
      kind: 'print',
      args: node.children.map(transformExpr),
      newline: callee === 'Serial.println',
    } as ASLStatement;
  }

  if (callee === 'Serial.begin') {
    return {
      kind: 'serialBegin',
      baud: transformExpr(node.children[0] || { nodeType: 'Literal', attributes: { value: 9600 }, children: [] } as any)
    } as ASLStatement;
  }

  if (callee === 'delayMicroseconds') {
    return {
      kind: 'expr',
      expr: {
        kind: 'call',
        callee: 'delayMicroseconds',
        args: node.children.map(transformExpr),
      },
    } as ASLStatement;
  }

  if (callee === 'Pin.value' && node.children.length === 2) {
    return {
      kind: 'digitalWrite',
      pin: transformExpr(node.children[0]),
      value: transformExpr(node.children[1]),
    } as ASLStatement;
  }

  if (callee === 'Pin.on') {
    return {
      kind: 'digitalWrite',
      pin: transformExpr(node.children[0]),
      value: { kind: 'literal', value: 1 } as ASLExpr,
    } as ASLStatement;
  }

  if (callee === 'Pin.off') {
    return {
      kind: 'digitalWrite',
      pin: transformExpr(node.children[0]),
      value: { kind: 'literal', value: 0 } as ASLExpr,
    } as ASLStatement;
  }

  // --- S5: UART high-level helpers ---
  if (callee === 'UARTWrite') {
    return {
      kind: 'uartWrite',
      port: transformExpr(node.children[0]),
      data: transformExpr(node.children[1]),
    } as ASLStatement;
  }

  if (callee === 'UARTRead') {
    return {
      kind: 'uartRead',
      port: transformExpr(node.children[0]),
      length: transformExpr(node.children[1]),
      target: node.attributes.target || '__uartBuf',
    } as ASLStatement;
  }

  // --- S5: I2C helpers ---
  if (callee === 'I2CWrite') {
    return {
      kind: 'i2cWrite',
      bus: transformExpr(node.children[0]),
      address: transformExpr(node.children[1]),
      data: transformExpr(node.children[2]),
    } as ASLStatement;
  }

  if (callee === 'I2CRead') {
    return {
      kind: 'i2cRead',
      bus: transformExpr(node.children[0]),
      address: transformExpr(node.children[1]),
      length: transformExpr(node.children[2]),
      target: node.attributes.target || '__i2cBuf',
    } as ASLStatement;
  }

  // --- S5: SPI helper ---
  if (callee === 'SPITransfer') {
    return {
      kind: 'spiTransfer',
      bus: transformExpr(node.children[0]),
      csPin: transformExpr(node.children[1]),
      txData: transformExpr(node.children[2]),
      target: node.attributes.target,
    } as ASLStatement;
  }

  // --- S5: PWM ---
  if (callee === 'PWMInit') {
    return {
      kind: 'pwmInit',
      pin: transformExpr(node.children[0]),
      freq: transformExpr(node.children[1]),
      duty: transformExpr(node.children[2]),
    } as ASLStatement;
  }
  if (callee === 'PWMSetDuty') {
    return {
      kind: 'pwmSetDuty',
      pin: transformExpr(node.children[0]),
      duty: transformExpr(node.children[1]),
    } as ASLStatement;
  }
  if (callee === 'PWMSetFreq') {
    return {
      kind: 'pwmSetFreq',
      pin: transformExpr(node.children[0]),
      freq: transformExpr(node.children[1]),
    } as ASLStatement;
  }
  if (callee === 'PWMStop') {
    return {
      kind: 'pwmStop',
      pin: transformExpr(node.children[0]),
    } as ASLStatement;
  }

  // --- S5: IEC blocks as function-like calls ---
  if (callee === 'TON') {
    return {
      kind: 'timerTON',
      instance: node.attributes.instance,
      in: transformExpr(node.children[0]),
      pt: transformExpr(node.children[1]),
    } as ASLStatement;
  }

  if (callee === 'TOF') {
    return {
      kind: 'timerTOF',
      instance: node.attributes.instance,
      in: transformExpr(node.children[0]),
      pt: transformExpr(node.children[1]),
    } as ASLStatement;
  }

  if (callee === 'TP') {
    return {
      kind: 'timerTP',
      instance: node.attributes.instance,
      in: transformExpr(node.children[0]),
      pt: transformExpr(node.children[1]),
    } as ASLStatement;
  }

  if (callee === 'CTU') {
    return {
      kind: 'counterCTU',
      instance: node.attributes.instance,
      cu: transformExpr(node.children[0]),
      r: transformExpr(node.children[1]),
      pv: transformExpr(node.children[2]),
    } as ASLStatement;
  }

  if (callee === 'CTD') {
    return {
      kind: 'counterCTD',
      instance: node.attributes.instance,
      cd: transformExpr(node.children[0]),
      ld: transformExpr(node.children[1]),
      pv: transformExpr(node.children[2]),
    } as ASLStatement;
  }

  if (callee === 'SR') {
    return {
      kind: 'latchSR',
      instance: node.attributes.instance,
      s: transformExpr(node.children[0]),
      r: transformExpr(node.children[1]),
    } as ASLStatement;
  }

  if (callee === 'RS') {
    return {
      kind: 'latchRS',
      instance: node.attributes.instance,
      r: transformExpr(node.children[0]),
      s: transformExpr(node.children[1]),
    } as ASLStatement;
  }

  if (callee === 'R_TRIG') {
    return {
      kind: 'trigR',
      instance: node.attributes.instance,
      in: transformExpr(node.children[0]),
    } as ASLStatement;
  }

  if (callee === 'F_TRIG') {
    return {
      kind: 'trigF',
      instance: node.attributes.instance,
      in: transformExpr(node.children[0]),
    } as ASLStatement;
  }

  if (callee === 'setColor' || (callee.endsWith && callee.endsWith('.setColor'))) {
    const objName = node.attributes.object || (callee.includes('.') ? callee.split('.')[0] : null);
    const pins = objName ? ctx.rgbPins?.get(objName) : null;
    if (pins) {
      const arg = node.children[0];
      let r: ASLExpr, g: ASLExpr, b: ASLExpr;

      if (arg && arg.nodeType === 'ArrayInitializer' && arg.children.length >= 3) {
        // setColor((255, 0, 0))
        r = transformExpr(arg.children[0]);
        g = transformExpr(arg.children[1]);
        b = transformExpr(arg.children[2]);
      } else if (node.children.length >= 3) {
        // setColor(255, 0, 0)
        r = transformExpr(node.children[0]);
        g = transformExpr(node.children[1]);
        b = transformExpr(node.children[2]);
      } else if (arg) {
        // setColor(c) -> c[0], c[1], c[2]
        const base = transformExpr(arg);
        r = { kind: 'index', target: base, index: { kind: 'literal', value: 0 } };
        g = { kind: 'index', target: base, index: { kind: 'literal', value: 1 } };
        b = { kind: 'index', target: base, index: { kind: 'literal', value: 2 } };
      } else {
        r = g = b = { kind: 'literal', value: 0 };
      }

      return {
        kind: 'rgbSet',
        pinR: transformExpr(pins.r),
        pinG: transformExpr(pins.g),
        pinB: transformExpr(pins.b),
        r, g, b
      } as ASLStatement;
    }
  }

  if (callee.endsWith && (callee === 'duty_u16' || callee.endsWith('.duty_u16'))) {
    const objName = callee.includes('.') ? callee.split('.')[0] : node.attributes.object;
    if (objName) {
      const pin = ctx.pwmPins?.get(objName);
      if (pin) {
        return {
          kind: 'analogWrite',
          pin: transformExpr(pin),
          value: {
            kind: 'binary',
            op: '/',
            left: transformExpr(node.children[0]),
            right: { kind: 'literal', value: 257 }
          }
        } as ASLStatement;
      }
    }
  }

  return {
    kind: 'expr',
    expr: {
      kind: 'call',
      callee,
      args: node.children.map(transformExpr),
    },
  } as ASLStatement;
}

export function tryTransformRead(targetVar: string, valueNode: BaseNode | undefined): ASLStatement | null {
  if (!valueNode) return null;

  if (valueNode.nodeType === 'GpioRead') {
    return {
      kind: 'read',
      mode: 'DIGITAL',
      target: targetVar,
      pin: transformExpr(valueNode.children[0]),
    } as ASLStatement;
  }
  if (valueNode.nodeType === 'AnalogRead') {
    return {
      kind: 'read',
      mode: 'ANALOG',
      target: targetVar,
      pin: transformExpr(valueNode.children[0]),
    } as ASLStatement;
  }
  if (valueNode.nodeType === 'CallExpression') {
    if (valueNode.attributes.callee === 'digitalRead') {
      return {
        kind: 'read',
        mode: 'DIGITAL',
        target: targetVar,
        pin: transformExpr(valueNode.children[0]),
      } as ASLStatement;
    }
    if (valueNode.attributes.callee === 'analogRead') {
      return {
        kind: 'read',
        mode: 'ANALOG',
        target: targetVar,
        pin: transformExpr(valueNode.children[0]),
      } as ASLStatement;
    }
    if (valueNode.attributes.callee === 'Pin.value') {
      return {
        kind: 'read',
        mode: 'DIGITAL',
        target: targetVar,
        pin: transformExpr(valueNode.children[0]),
      } as ASLStatement;
    }
  }
  return null;
}

File: crates/neuroforge-asl/src/transforms/call_transform.rs
//! call_transform.rs — transforma chamadas de função em nós ASL de hardware
//! Migrado de: src/engine/asl/transforms/callTransform.ts

use crate::transforms::context::TransformContext;
use crate::transforms::expr_transform::transform_expr;
use crate::types::asl_types::*;
use crate::types::typed_nodes::CallNode;

/// Resultado de transformar uma chamada de função.
/// Pode produzir um AslStatement (hardware call) ou nada (chamada desconhecida).
pub fn transform_call(call: &CallNode, ctx: &TransformContext) -> Option<AslStatement> {
    let args: Vec<AslExpr> = call.args.iter().map(|a| transform_expr(a, ctx)).collect();

    match call.name.as_str() {
        // ── GPIO ─────────────────────────────────────────────────────────────
        "pinMode" => {
            let pin = args.first().cloned().unwrap_or(AslExpr::int(0));
            let mode_arg = call.args.get(1);
            let is_output = match mode_arg {
                Some(arg) => match &arg.kind {
                    crate::types::typed_nodes::ExprKind::StringLiteral(s) => s == "OUTPUT",
                    crate::types::typed_nodes::ExprKind::Identifier(s) => s == "OUTPUT",
                    _ => false,
                },
                None => true,
            };
            let mode = if is_output {
                PinModeKind::Output
            } else {
                PinModeKind::Input
            };
            Some(AslStatement::PinMode(AslPinMode { pin, mode }))
        }
        "digitalWrite" => {
            let pin = args.first().cloned().unwrap_or(AslExpr::int(0));
            // Check if arg is HIGH/LOW/1/0
            let value_arg = call.args.get(1);
            let value = match value_arg {
                Some(arg) => match &arg.kind {
                    crate::types::typed_nodes::ExprKind::Identifier(s) => {
                        if s == "HIGH" { DigitalValue::High }
                        else if s == "LOW" { DigitalValue::Low }
                        else { DigitalValue::Expr(args.get(1).cloned().unwrap_or(AslExpr::int(0))) }
                    },
                    crate::types::typed_nodes::ExprKind::IntLiteral(n) => {
                        if *n > 0 { DigitalValue::High }
                        else { DigitalValue::Low }
                    },
                    _ => DigitalValue::Expr(args.get(1).cloned().unwrap_or(AslExpr::int(0))),
                },
                None => DigitalValue::Low,
            };
            Some(AslStatement::DigitalWrite(AslDigitalWrite { pin, value }))
        }
        "analogWrite" => {
            let pin = args.first().cloned().unwrap_or(AslExpr::int(0));
            let value = args.get(1).cloned().unwrap_or(AslExpr::int(0));
            Some(AslStatement::AnalogWrite(AslAnalogWrite { pin, value }))
        }
        "digitalRead" | "analogRead" => {
            let pin = args.first().cloned().unwrap_or(AslExpr::int(0));
            let mode = if call.name == "analogRead" {
                ReadMode::Analog
            } else {
                ReadMode::Digital
            };
            let result = call.result_var.clone().unwrap_or_default();
            Some(AslStatement::Read(AslRead {
                pin,
                mode,
                target: result,
            }))
        }

        // ── Serial ───────────────────────────────────────────────────────────
        "Serial.begin" | "serialBegin" => {
            let baud = args.first().cloned().unwrap_or(AslExpr::int(9600));
            Some(AslStatement::SerialBegin(AslSerialBegin { baud }))
        }
        "Serial.print" | "Serial.println" | "print" => {
            let value = args.first().cloned().unwrap_or(AslExpr::str_val(""));
            let newline = call.name.ends_with("println");
            Some(AslStatement::Print(AslPrint {
                args: vec![value],
                newline,
            }))
        }

        // ── Timing ───────────────────────────────────────────────────────────
        "delay" => {
            let ms = args.first().cloned().unwrap_or(AslExpr::int(0));
            Some(AslStatement::Delay(AslDelay { milliseconds: ms }))
        }
        "delayMicroseconds" => {
            let us = args.first().cloned().unwrap_or(AslExpr::int(0));
            Some(AslStatement::Delay(AslDelay { milliseconds: us }))
        }

        // ── Servo ────────────────────────────────────────────────────────────
        "servo.attach" | "servoAttach" => {
            let pin = args.first().cloned().unwrap_or(AslExpr::int(0));
            let name = call.object.clone().unwrap_or_else(|| "servo".into());
            Some(AslStatement::ServoAttach(AslServoAttach {
                var_name: name,
                pin,
                min_pulse: None,
                max_pulse: None,
                continuous: None,
            }))
        }
        "servo.write" | "servoWrite" => {
            let angle = args.first().cloned().unwrap_or(AslExpr::int(0));
            let name = call.object.clone().unwrap_or_else(|| "servo".into());
            Some(AslStatement::ServoWrite(AslServoWrite {
                var_name: name,
                angle,
                raw_microseconds: None,
            }))
        }
        "servo.detach" | "servoDetach" => {
            let name = call.object.clone().unwrap_or_else(|| "servo".into());
            Some(AslStatement::ServoDetach(AslServoDetach { var_name: name }))
        }

        // ── I2C ──────────────────────────────────────────────────────────────
        "Wire.write" | "i2cWrite" => {
            let addr = args.first().cloned().unwrap_or(AslExpr::int(0));
            let value = args.get(1).cloned().unwrap_or(AslExpr::int(0));
            Some(AslStatement::I2cWrite(AslI2cWrite {
                bus: AslExpr::int(0),
                address: addr,
                data: value,
            }))
        }
        "Wire.read" | "i2cRead" => {
            let addr = args.first().cloned().unwrap_or(AslExpr::int(0));
            let result = call.result_var.clone().unwrap_or_default();
            Some(AslStatement::I2cRead(AslI2cRead {
                bus: AslExpr::int(0),
                address: addr,
                target: result,
                length: AslExpr::int(1),
            }))
        }

        // ── SPI ──────────────────────────────────────────────────────────────
        "SPI.transfer" | "spiTransfer" => {
            let value = args.first().cloned().unwrap_or(AslExpr::int(0));
            let result = call.result_var.clone().unwrap_or_default();
            Some(AslStatement::SpiTransfer(AslSpiTransfer {
                bus: AslExpr::int(0),
                cs_pin: AslExpr::int(0),
                tx_data: value,
                target: Some(result),
            }))
        }

        // ── PWM ──────────────────────────────────────────────────────────────
        "pwmInit" => {
            let pin = args.first().cloned().unwrap_or(AslExpr::int(0));
            let freq = args.get(1).cloned().unwrap_or(AslExpr::int(1000));
            Some(AslStatement::PwmInit(AslPwmInit {
                pin,
                freq,
                duty: AslExpr::int(0),
            }))
        }
        "pwmSetDuty" => {
            let pin = args.first().cloned().unwrap_or(AslExpr::int(0));
            let duty = args.get(1).cloned().unwrap_or(AslExpr::int(0));
            Some(AslStatement::PwmSetDuty(AslPwmSetDuty { pin, duty }))
        }

        // ── RGB ──────────────────────────────────────────────────────────────
        "rgbSet" => {
            let pin = args.first().cloned().unwrap_or(AslExpr::int(0));
            let r = args.get(1).cloned().unwrap_or(AslExpr::int(0));
            let g = args.get(2).cloned().unwrap_or(AslExpr::int(0));
            let _b = args.get(3).cloned().unwrap_or(AslExpr::int(0));
            Some(AslStatement::RgbSet(AslRgbSet {
                pin_r: pin,
                pin_g: r,
                pin_b: g,
                r: AslExpr::int(0),
                g: AslExpr::int(0),
                b: AslExpr::int(0),
            }))
        }

        // ── Chamada genérica (função de utilizador) ──────────────────────────
        _ => Some(AslStatement::Expr(AslExpressionStmt {
            expr: AslExpr::Call(Box::new(AslCall {
                callee: call.name.clone(),
                args,
            })),
        })),
    }
}
